### ISSUE TITLE
Bjerksund-Stensland analytic approximation for American options

### ISSUE THREAD

**Issue by cyrilchim** [google/tf-quant-finance#5]
[Bjerksund-Stensland](https://www.sciencedirect.com/science/article/pii/095652219390009H) approximation is an alternative to the [Baron-Adesi Whaley](https://deriscope.com/docs/Barone_Adesi_Whaley_1987.pdf) approximation for  American options pricing formula under Black-Scholes. 

The module implementing this method should live under tf_quant_finance/volatility/american_option.py. It should support both puts and calls. Tests should be in american_option_test.py in the same folder.

**Comment by 2torus** [comment]
@cyrilchim I'd like to work on this.

**Comment by saxena-ashish-g** [comment]
@2torus: sounds good. I will assign it to you. Please contact us if you have any questions.

**Comment by DevarakondaV** [comment]
Is @2torus still working on this?

**Comment by saxena-ashish-g** [comment]
> Is @2torus still working on this?

We haven't seen any activity on this issue since September. If @2torus isn't, would you like to tackle this one @DevarakondaV ?

**Comment by DevarakondaV** [comment]
Yeah!

**Comment by cyrilchim** [comment]
Thanks!

**Comment by 2torus** [comment]
@cyrilchim Wait, I did half the work. https://github.com/2torus/tf-quant-finance/blob/BjerksunStensland/tf_quant_finance/black_scholes/approximations/bjerksund_stensland.py Give me a couple of weeks to wrap it up, please.

**Comment by cyrilchim** [comment]
Oh, I am sorry Boris. Reassigning back to @2torus

**Comment by 2torus** [comment]
It's totally my fault. I was doing it at a good pace until life interfered :-(

**Comment by cyrilchim** [comment]
@DevarakondaV  Would you like to get a crack on [barrier options](https://github.com/google/tf-quant-finance/issues/7)? We really need that one but there was no progress on that front so far?

**Comment by DevarakondaV** [comment]
> @DevarakondaV Would you like to get a crack on [barrier options](https://github.com/google/tf-quant-finance/issues/7)? We really need that one but there was no progress on that front so far?

I'll have to familiarize myself with Barrier Options, but I'm guessing I should be able to do it. It might take some time though.

**Comment by alembcke** [comment]
I haven't seen any updates on this in a while, so I thought I would give it a go.  But I have run into an issue I cannot resolve.

The Bjerksund-Stensland 2002 algorithm utilizes the Standard Bivariate Normal Distribution, specifically the Cumulative Distribution Function (CDF) of that distribution.  While TensorFlow Probability has the class `tfp.distributions.MultivariateNormalTriL`, the CDF function is not implemented (see [TFP #1361](https://github.com/tensorflow/probability/issues/1361)).

The only solution I have found is to use `scipy.stats.multivariate_normal`, but that would add a new dependency to the library (from what I can see).  Does anyone else have any ideas or suggestions?

**Comment by cyrilchim** [comment]
Hi Alex, 

Indeed, the CDF is not implemented. The best way to go is to have a utility function for that. [Here is C code](https://people.sc.fsu.edu/~jburkardt/c_src/toms462/toms462.c) for computing  the bivariate normal cdf (you need `bivnor` function). It should be easy to translate that to Python but might take some effort making it TF-friendly. Please let me know if you need help implementing the utility.

**Comment by cyrilchim** [comment]
Reference paper is [this one](https://dl.acm.org/doi/abs/10.1145/362375.362414)

**Comment by cyrilchim** [comment]
Here is probably a better [reference](https://projecteuclid.org/journals/annals-of-mathematical-statistics/volume-27/issue-4/Tables-for-Computing-Bivariate-Normal-Probabilities/10.1214/aoms/1177728074.full) I think this paper closely follows the logic from the code

**Comment by alembcke** [comment]
In Haug's book The Complete Guide to Option Pricing Formulas, he mentions the Genz (2004) algorithm as the most accurate (reference paper is [here](http://www.math.wsu.edu/faculty/genz/papers/bvnt.pdf)) and provides VBA code.  And Genz's website [here](http://www.math.wsu.edu/faculty/genz/homepage) provides Fortran and Matlab code.  Any objections to going with this one instead?

**Comment by cyrilchim** [comment]
Thanks, Alex! Genz's approach seems like the right way to go.

**Comment by alembcke** [comment]
@cyrilchim I have made a lot of progress on implementing Genz's approach for the standard bivariate normal distribution and the Bjerksund-Stensland 2002 algorithm.  Here is the code as it stands at the moment:  [link](https://github.com/alembcke/tf-quant-finance/blob/bs2002_tensor/tf_quant_finance/black_scholes/approximations/american_option.py)

The code works in eager mode, and the values are almost matching the reference paper.  But the real issue I am facing, where I could use the advice of a seasoned TensorFlow developer (I am relatively new to the TF world), is that Genz's algorithm has lots of `if else` statements and `for` loops.  What is the best way to tensorify these?

I started down the path of converting all of the `if else` statements to `tf.cond()` statements, but the real issue is that the loops have a variable upper bound (via the variables `ng` and `lg`).  What is the best way to handle these?

Also, could you assign this issue to me, as I am actively working on it and close to a solution?  I hope you don't mind @2torus .

**Comment by 2torus** [comment]
@alembcke No problem. Great work so far. 
Try `tf.while_loop` [docs](https://www.tensorflow.org/api_docs/python/tf/while_loop) for your loops. 
I also have a notebook on how to compute this approximation with QuantLib that I could share.

**Comment by cyrilchim** [comment]
@2torus, please do share if you could. Thank you!

@alembcke Well done on the progress! You could use `tf.cond` and put it inside `tf.vectorized_map` over all entries of `rho`. It might be not that great for GPUs though. So we typically use `tf.where` instead. It is best to vectorized the code as much as you can (please let me know if you need help with this step).
 
As for the for-loop, you could use `tf.while_loop` as @2torus recommended above. I think you should be able to vectorize this calculation explicitly instead:

 For example, instead of 

```python
          for i in range(0, lg):
              for is1 in range(-1, 3, 2):
                  sn = np.sin(asr * (is1 * x[ng, i] + 1) / 2)
                  bvn = bvn + w[ng, i] * np.exp((sn * hk - hs) / (1 - sn * sn))
```

you could try something like

```python
         # shape `[2, 1]`
         signs = tf.constant([[-1], [1]], dtype=x.dtype)
        # Assume shape of `asr` is `batch_shape`. Then asr[..., None, None] is of shape `batch_shape + [1, 1]`
        # shape` batch_shape + [2, lg]`
         sn = np.sin(asr[..., None, None] * (is1 * x[ng] + 1) / 2)
        # shape `batch_shape`
         bvn += tf.reduce_sum(w[ng] * np.exp((sn * hk - hs) / (1 - sn * sn)), axis=[-2, -1])
``` 


P.S. if you are just writing code for yourself, you could try wrapping the function with `tf.function(autograph=True)` and see if [autograph](https://www.tensorflow.org/guide/function#autograph_transformations) can make the code tf-friendly

**Comment by 2torus** [comment]
@cyrilchim A working colab with examples from quantlib https://colab.research.google.com/drive/1bEFOfI6y8lulViqsnodTT4wKt4ZgjuOV?usp=sharing

**Comment by alembcke** [comment]
Thank you @2torus , this will certainly come in handy.

@cyrilchim , are there any good resources you could recommend to learn about vectorizing code?  The code I have written for the 2002 algorithm is rather slow, so speeding it up while adapting it to work in graph mode sounds like a good idea.

**Comment by 2torus** [comment]
@alembcke I can lend a hand with vectorization, if you want to.

**Comment by cyrilchim** [comment]
@alembcke I am not sure if there is a good reference on vectorization. Boris, if you could help with that, that would be much appreciated. As for performance, typically JIT compilation is important. You would need to use `tf.function(fn, jit_compile=True)` (see, e.g., [here](https://www.tensorflow.org/xla/tutorials/jit_compile)).

**Comment by alembcke** [comment]
@2torus your help would be appreciated.  I will try to do this myself as well, as it is a good learning experience for me.  But I'm sure I can also learn a lot by examining your vectorized code and I have other projects which will give me the opportunity to apply this technique.

@cyrilchim  searching the TensorFlow website, I came across this paper [link](https://arxiv.org/pdf/1903.04243.pdf).  It seems to be a good place to start.

**Comment by 2torus** [comment]
@alembcke Just show the work in progress.

**Comment by alembcke** [comment]
@2torus here is my attempt to tensorfy the BVND function:  [link](https://github.com/alembcke/tf-quant-finance/blob/bs2002/tf_quant_finance/black_scholes/approximations/american_option.py)

And here is the python code, prior to my attempt to tensorfy:  [link](https://github.com/alembcke/tf-quant-finance/blob/bs2002_tensor/tf_quant_finance/black_scholes/approximations/american_option.py)

Both run in eager mode, but neither works in graph mode.  Let me know if there is anything else you need.

**Comment by cyrilchim** [comment]
@alembcke I have update TF trainings a bit. There is a section on Autograph,` tf.while_loop`, and `tf.cond` in [second training](https://colab.research.google.com/github/google/tf-quant-finance/blob/master/tf_quant_finance/examples/jupyter_notebooks/Introduction_to_TensorFlow_Part_2_-_Debugging_and_Control_Flow.ipynb) 

Looking at the code (BVDN), you should make it work in graph mode and will probably be fast with XLA-compilation (see the [third training](https://colab.research.google.com/github/google/tf-quant-finance/blob/master/tf_quant_finance/examples/jupyter_notebooks/Introduction_to_TensorFlow_Part_3_-_Advanced_Tensor_Manipulation.ipynb) on that). 

It would be nice to have the code run well for the GPU as well but conditional branching will be quite detrimental. I have not read through the paper carefully but it feels like branching is done because an integral is approximated with a different number of points depending on the value of `rho`. Could you please confirm this is the case? If so, I would suggest we simply use the same number of points for all values of rho (the largest number). This might work slightly worse on a CPU but due to SIMD threading might not be that bad when pricing a batch of options. Also, this should be well suited for GPUs. 

Please let me know if this makes sense.

**Comment by alembcke** [comment]
@cyrilchim thank you for updating the TF trainings.  I went through them when I first started on this, but wasn't able to put a lot of it in context since I was so new to TF, so am going through them again.

I updated the code to use `tf.while_loop` and some other adjustments to work in graph mode [link](https://github.com/alembcke/tf-quant-finance/blob/bs2002/tf_quant_finance/black_scholes/approximations/american_option.py).  So at least we have code that works in both eager and graph modes now.  So here is what seems to be outstanding to me:

## Improvements to speed up computation time
The code takes quite a while to run and you have mentioned some areas that could help speed up the computation time, such as:
1. Vectorizing the code.
2. Reviewing the BVND paper to see if we can reduce the conditional branching.

Just to quantify the speed I ran the various american option pricing models in eager mode for a set of 5 options and here are the results:
```
CRR Binomial (num_steps=100) elapsed 0.078 seconds.
Barone-Adesi Whaley elapsed 0.036 seconds.
Bjerksund-Stensland 1993 elapsed 0.032 seconds.
Bjerksund-Stensland 2002 elapsed 1.092 seconds.
```
This test is just for comparison purposes and is not meant to be comprehensive, but shows a clear difference.

## The values returned can be off by as much as 0.02
This one I am stumped on.  I have gone through the calculations at least 10 times and cannot find the issue.  I was wondering if @2torus Colab notebook using QuantLib had the same deviations, but QuantLib only supports the 1993 algorithm as per [link](http://www.quantlib.org/reference/class_quant_lib_1_1_bjerksund_stensland_approximation_engine.html).  Here are the results of [this](https://github.com/alembcke/tf_quant_finance_testing/blob/main/bs2002_test64.py) test program, which illustrate the deviations:
```
Closed Form Valuation of American Options, 2002
Table 1, pg. 9 - 2002 model
Section 1, Calls:  [ 0.03  0.57  3.5  10.33 20.  ]  Puts:  [20.41 11.25  4.4   1.12  0.18]
 Expected Values   [ 0.03  0.58  3.51 10.34 20.00]         [20.41 11.25  4.40  1.12  0.18]
Section 2, Calls:  [ 0.03  0.57  3.48 10.31 20.  ]  Puts:  [20.22 11.14  4.35  1.11  0.18]
 Expected Values   [ 0.03  0.57  3.49 10.31 20.00]         [20.23 11.14  4.35  1.11  0.18]
Section 3, Calls:  [ 1.05  3.26  7.39 13.5  21.25]  Puts:  [21.44 13.91  8.27  4.52  2.29]
 Expected Values   [ 1.05  3.26  7.39 13.51 21.26]         [21.44 13.91  8.27  4.52  2.29]
Section 4, Calls:  [ 0.21  1.34  4.67 10.96 20.  ]  Puts:  [20.95 12.63  6.37  2.65  0.92]
 Expected Values   [ 0.21  1.35  4.69 10.98 20.00]         [20.96 12.63  6.37  2.65  0.92]
Table 2, pg. 10 - 2002 model
Section 1, Calls:  [ 0.05  0.85  4.44 11.66 20.9 ]  Puts:  [20.   10.2   3.52  0.78  0.11]
 Expected Values   [ 0.05  0.85  4.44 11.66 20.90]         [20.00 10.21  3.53  0.79  0.11]
Section 2, Calls:  [ 0.05  0.84  4.4  11.55 20.69]  Puts:  [20.   10.18  3.5   0.78  0.11]
 Expected Values   [ 0.05  0.84  4.40 11.55 20.69]         [20.00 10.19  3.51  0.78  0.11]
Section 3, Calls:  [ 1.29  3.82  8.35 14.8  22.71]  Puts:  [20.54 12.93  7.44  3.94  1.94]
 Expected Values   [ 1.29  3.82  8.35 14.80 22.71]         [20.55 12.94  7.45  3.94  1.94]
Section 4, Calls:  [ 0.41  2.18  6.5  13.42 22.06]  Puts:  [20.   10.72  4.73  1.72  0.52]
 Expected Values   [ 0.41  2.18  6.50 13.42 20.06]         [20.00 10.73  4.74  1.72  0.52]
Table 3, pg. 11 - 2002 model
Section 1,                                          Puts:  [20.   10.02  3.18  0.65  0.09]
 Expected Values                                           [20.00 10.02  3.20  0.66  0.09]
Section 2,                                          Puts:  [20.   10.    2.88  0.54  0.07]
 Expected Values                                           [20.00 10.00  2.90  0.55  0.07]
Section 3,                                          Puts:  [20.29 12.52  7.07  3.67  1.78]
 Expected Values                                           [20.30 12.54  7.09  3.69  1.78]
Section 4,                                          Puts:  [20.   10.25  4.13  1.38  0.39]
 Expected Values                                           [20.00 10.27  4.15  1.39  0.39]
Table 4, pg. 12 - 2002 model
Section 1, Calls:  [ 2.32  4.72  8.45 13.76 20.86]  Puts:  [25.63 20.06 15.48 11.79  8.88]
 Expected Values   [ 2.32  4.74  8.47 13.77 20.86]         [25.64 20.07 15.49 11.80  8.88]
Section 2, Calls:  [ 3.97  7.23 11.67 17.27 23.95]  Puts:  [22.13 16.17 11.67  8.34  5.91]
 Expected Values   [ 3.97  7.23 11.68 17.28 23.95]         [22.14 16.17 11.68  8.35  5.91]
Section 3, Calls:  [ 6.88 11.49 17.21 23.84 31.16]  Puts:  [20.32 13.45  8.89  5.86  3.85]
 Expected Values   [ 6.88 11.49 17.21 23.84 31.16]         [20.33 13.47  8.91  5.88  3.87]
Section 4,                                          Puts:  [20.   11.67  6.9   4.13  2.48]
 Expected Values                                           [20.00 11.68  6.91  4.13  2.49]
The Complete Guide to Option Pricing Formulas, 2007
Table 3-2, pg. 107, 2002 model
Call Options:
T=0.1 sigma=0.15:  [ 0.0205  1.8757 10.    ]
      Expected:    [ 0.0205  1.8757 10.0000]
      sigma=0.25:  [ 0.3151  3.1256 10.3725]
      Expected:    [ 0.3151  3.1256 10.3725]
      sigma=0.35:  [ 0.9479  4.3746 11.1578]
      Expected:    [ 0.9479  4.3746 11.1578]
T=0.5 sigma=0.15:  [ 0.8099  4.0628 10.7898]
      Expected:    [ 0.8099  4.0628 10.7898]
      sigma=0.25:  [ 2.718   6.7661 12.9814]
      Expected:    [ 2.7180  6.7661 12.9814]
      sigma=0.35:  [ 4.9665  9.4608 15.5137]
      Expected:    [ 4.9665  9.4608 15.5137]
Put Options:
T=0.1 sigma=0.15:  [10.      1.8757  0.0408]
      Expected:    [10.0000  1.8757  0.0408]
      sigma=0.25:  [10.228   3.1256  0.4552]
      Expected:    [10.2280  3.1256  0.4552]
      sigma=0.35:  [10.8663  4.3746  1.2383]
      Expected:    [10.8663  4.3746  1.2383]
T=0.5 sigma=0.15:  [10.54    4.0628  1.0689]
      Expected:    [10.5400  4.0628  1.0689]
      sigma=0.25:  [12.4097  6.7661  3.2932]
      Expected:    [12.4097  6.7661  3.2932]
      sigma=0.35:  [14.6445  9.4608  5.8374]
      Expected:    [14.6445  9.4608  5.8374]
```

I am going to keep investigating to see if I can find the issue for the slight deviations, but let me know if you spot anything.

Those are the open issues as far as I can see, but please let me know if I missed anything.  I didn't mention the price deviations before, as was focused on getting the code working in both eager and graph mode as a first step.  Now I will focus on finding the reason for the deviations and then I will move on to the computational improvements.

**Comment by cyrilchim** [comment]
Thank you, @alembcke! This looks great. I have tweaked a few things in [this colab](https://colab.research.google.com/drive/1qXrmVgGV4BekAlriTmt5k-flJX-E9r6G?resourcekey=0-Zc2lOkoW8dvkoGewxWmkoA&usp=sharing) (let me know if you do not have access). A few notes:

1. Let us use 10 points for Legendre integration regardless of `rho` values
2. I think your benchmarks are for eager mode, where Python overhead dominated. I have added benchmarks in graph mode with XLA compilation and got `800 micro-secs` for 2002 algorithm and `500 micro-secs` for 1993 algorithm. Now, this is on a CPU (the benchmarks are in the colab). 
3. GPU performance is really bad for 2002 algorithm (`22 ms`). I am sure that `tf.cond` is to blame. Since I think we should always use 10 interpolation points, many of `tf.cond` can be removed.

Please check out [this](https://colab.research.google.com/github/google/tf-quant-finance/blob/master/tf_quant_finance/examples/jupyter_notebooks/Vectorization_and_XLA_compilation.ipynb) colab for vectorization and XLA compilation demo. 

Please let me know if this makes sense.

**Comment by alembcke** [comment]
@cyrilchim I went through Genz's paper and the code to determine if there are any branches we can get rid of.  Turns out `tf.math.abs(rho)` is always the same value `tf.math.abs(rho) = tf.math.sqrt(0.5 * (tf.math.sqrt(5.0) - 1))` which is roughly `0.78615`.  So the only branch executed is `if_a1` which always executes `if_b1`.  So all we need is this:

```py
def _cbnd(dh, dk, rho):
  """Compute values for the cumulative standard bivariate normal distribution.
  #### References:
  [1] Genz, A., Numerical Computation of Rectanglar Bivariate and Trivariate
        Normal and t Probabilities, 2004
    http://www.math.wsu.edu/faculty/genz/papers/bvnt.pdf"""
  
  # Python code adapted from Fortran code and naming convention will align
  # variables with the variables named in BVND function, except `IS` becomes
  # `is1` and `AS` becomes `as1` since `is` and `as` are Python keywords.
  # http://www.math.wsu.edu/faculty/genz/software/fort77/tvpack.f
 
  dtype = rho.dtype

  # Gauss Legendre Points and Weights
  w = tf.constant([0.01761400713915212, 0.04060142980038694, 0.06267204833410906,
                   0.08327674157670475, 0.1019301198172404, 0.1181945319615184,
                   0.1316886384491766, 0.1420961093183821, 0.1491729864726037,
                   0.1527533871307259], dtype=dtype)

  x = tf.constant([-0.9931285991850949, -0.9639719272779138, -0.9122344282513259,
                   -0.8391169718222188, -0.7463319064601508, -0.6360536807265150,
                   -0.5108670019508271, -0.3737060887154196, -0.2277858511416451,
                   -0.07652652113349733], dtype=dtype)

  h = tf.cast(-dh, dtype=dtype)
  k = tf.cast(-dk, dtype=dtype)
  hk = h * k

  # In order to avoid `shape_invariants` argument in the while_loop, set
  # `bvn` to be a Tensor of zeros of the appropriate shape
  bvn = tf.zeros_like(hk)                                                                  
  
  hs = (h * h + k * k) / 2
  asr = tf.math.asin(rho)    
                                       
  def loop_b(bvn, is1, i, hk, hs, asr, x, w):    
      sn = tf.math.sin(asr * (is1 * x[i] + 1) / 2)    
      res = bvn + w[i] * tf.math.exp((sn * hk - hs) / (1 - sn * sn))    
      return res    

  def loop_a(bvn, i, hk, hs, asr, x, w):    
      is1 = tf.constant(-1, dtype=dtype)    
      is1, bvn = tf.while_loop(    
              loop_vars=[is1, bvn],    
              cond = lambda is1, bvn: is1 < 2,    
              body = lambda is1, bvn: [is1 + 2, loop_b(bvn, is1, i, hk, hs, asr, x, w)]    
              )
  i = tf.constant(0, dtype=tf.int32)
  i, bvn = tf.while_loop(
          loop_vars=[i, bvn],
          cond = lambda i, bvn: i < tf.size(x),
          body = lambda i, bvn: [i + 1, loop_a(bvn, i, hk, hs, asr, x, w)]
          )

  bvn = bvn * asr / (4 * np.pi)
  bvn = bvn + _ncdf(-h) * _ncdf(-k)

  return bvn
```

With this change, the 2002 algorithm drops below `600 micro-secs` with XLA compilation on the CPU.

I made this change to the code on my PC and all the tests complete successfully.  Do you think there are any further performance enhancements which can be done, or should I open a pull request with this code?

**Comment by cyrilchim** [comment]
Thank you, Alex. There is one more possible optimization. You could parallelize your loops. You have two options: either batch the calculation or use `tf.vectorized_map`. I have updated our [colab](https://colab.research.google.com/drive/1qXrmVgGV4BekAlriTmt5k-flJX-E9r6G?resourcekey=0-Zc2lOkoW8dvkoGewxWmkoA&usp=sharing) with a batched version (function called `batched_calc`). Please check that this passes the tests. It does not make much of a difference for a CPU run (when XLA-compiled), but definitely improves GPU performance (100k options are priced in `70ms` with 2002 algorithm and in `10ms` with 1193 algorithm)

**Comment by alembcke** [comment]
Colab ran fine with `batched_calc`, but then on my PC both eager and deferred (graph) mode complained about the shape of tensors on the line calculating `res`.  The solution was to change `for _ in range(asr.shape.rank):` to `for _ in range(hk.shape.rank):`.

The reason being that the shape of `asr` is based on the shape of `rho` and `rho` is based on the shape of `expiries`.  But the tensor which defines the shape of other tensors is `volatilities`, for which the shape of `hk` is based on.  Case in point, if you have five `spots` with one `expiries` then `expiries.shape = 1` when `volatilities.shape = 5`.  This wasn't a problem in the colab, as the shape of `expiries` matched the shape of other tensors.

I hope that makes sense and you agree.

Anyways, after making the change the colab ran successfully, as did eager and deferred (graph) mode tests on my PC.  So I have created pull request #59.  Let me know if it looks good or if I need to make any other changes.

And @cyrilchim  thank you for all your help, I have learned a lot working on this project with you.  I appreciate your patience and guidance throughout.
